pkg/operator/controller/ingress/controller.go

```diff
@@ -120,9 +124,182 @@ func New(mgr manager.Manager, config Config) (controller.Controller, error) {
 	if err := c.Watch(&source.Kind{Type: &configv1.Ingress{}}, handler.EnqueueRequestsFromMapFunc(reconciler.ingressConfigToIngressController)); err != nil {
 		return nil, err
 	}
+	// add watch for changes in Route
+	if err := c.Watch(&source.Kind{Type: &routev1.Route{}}, handler.EnqueueRequestsFromMapFunc(reconciler.gatherRouteMetrics)); err != nil {
+		return nil, err
+	}
 	return c, nil
 }
 
+// gatherRouteMetrics processes a Route object and gathers the IngressControllerRouteType metric
+// and IngressControllerRoutesPerShard metric.
+func (r *reconciler) gatherRouteMetrics(o client.Object) []reconcile.Request {
+	// Processing every Route event sequentially while gathering Route Metrics.
+	r.lock.Lock()
```

Comment:
Why do you need to lock the entire function?  That is probably too long for a lock to be held.